torch/testing/_internal/optests/aot_autograd.py

```diff
@@ -0,0 +1,112 @@
+import torch
+import torch.utils._pytree as pytree
+from torch.testing._internal.common_methods_invocations import wrapper_set_seed
+from functorch.compile import compiled_function, min_cut_rematerialization_partition, nop
+import re
+
+
+class assert_raises_regex:
+    def __init__(self, exception_cls, regex):
+        self.exception_cls = exception_cls
+        self.regex = regex
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, exc_type, exc_val, traceback):
+        if exc_type == self.exception_cls:
+            msg = str(exc_val)
+            if not re.search(self.regex, msg):
+                raise AssertionError(
+                    f"Expected exception to match regex. regex: {self.regex}, exception: {msg}")
+            return True  # Squashes the exception
+        if exc_type is not None:
+            raise AssertionError(
+                f"Expected {self.exception_cls} to be raised, instead got exception {exc_type}")
+        raise AssertionError("Expected exception to be raised but none was")
+
+
+def aot_autograd_check(
+        func, args, kwargs, dynamic,
+        assert_raises_regex_fn=assert_raises_regex,
+        assert_equals_fn=torch.testing._comparison.assert_close):
+    flat_args, args_spec = pytree.tree_flatten((args, kwargs))
+    for arg in flat_args:
+        if isinstance(arg, int) and arg == -42:
+            raise TestFrameworkError("We've reserved -42 as an arg")
+
+    sentinel_val = -42
+    is_tensor_spec = [sentinel_val if isinstance(arg, torch.Tensor) else arg for arg in flat_args]
+    args = [arg for arg in flat_args if isinstance(arg, torch.Tensor)]
+
+    def f(args):
+        cur_flat_args = list(is_tensor_spec)
+        args = iter(args)
+        for idx, v in enumerate(cur_flat_args):
+            if v == sentinel_val:
+                cur_flat_args[idx] = next(args)
+        c_args, c_kwargs = pytree.tree_unflatten(cur_flat_args, args_spec)
+        return func(*c_args, **c_kwargs)
+
+    compiled_f = compiled_function(f, nop, nop, dynamic=dynamic, partition_fn=min_cut_rematerialization_partition)
+    _test_aot_autograd_forwards_backwards_helper(
```

Comment:
This test will automatically decide to run aot autograd for inference vs training based on the gradness of the inputs. This is more of a UX question: do you think we should leave it on the user to decide to pass in requires_grad vs inference tensors when running `custom_op_check`? The other option is that we run the test twice, once with inference tensors and once without, regardless of the gradness of the tensors / current grad mode state of when from the user.

Reply:
We need to leave it to the user, because we don't know which Tensors to actually set requires_grad = True on. For example, if the user were testing batch_norm, then it would be bad if we automatically set requires_grad=True on the buffers.

Reply:
ah fair point